modern/org/commcare/modern/database/DatabaseHelper.java

```diff
@@ -0,0 +1,151 @@
+            if(fields != null) {
+                if(!fields.contains(columnName)) {
+                    continue;
+                }
+            }
+
+            if(set){
+                stringBuilder.append(" AND ");
+            }
+
+            stringBuilder.append(columnName);
+            stringBuilder.append("=?");
+
+            arguments.add(values[i].toString());
+
+            set = true;
+        }
+
+        String[] retArray = new String[arguments.size()];
+        for(int i =0; i< arguments.size(); i++){
+            retArray[i] = arguments.get(i);
+        }
+
+        return new Pair<String, String[]>(stringBuilder.toString(), retArray);
+    }
+
+    public static Set<String> getMetaDataFields(Persistable p){
+        Set<String> fields = null;
+        if(p instanceof IMetaData) {
+            IMetaData m = (IMetaData)p;
+            String[] thefields = m.getMetaDataFields();
+            fields = new HashSet<String>();
+            for(String s : thefields) {
+                fields.add(TableBuilder.scrubName(s));
+            }
+        }
+        return fields;
+    }
+
+
+    public static HashMap<String, Object> getMetaFieldsAndValues(Externalizable e) throws RecordTooLargeException{
+
+        ByteArrayOutputStream bos = new ByteArrayOutputStream();
+
+        try {
+            e.writeExternal(new DataOutputStream(bos));
+            bos.close();
+        } catch (IOException e1) {
+            e1.printStackTrace();
+            throw new RuntimeException("Failed to serialize externalizable for content values");
+        }
+        byte[] blob = bos.toByteArray();
+
+        HashMap<String, Object> values = new HashMap<String, Object>();
+
+        if(e instanceof IMetaData) {
+            IMetaData m = (IMetaData)e;
+            for(String key : m.getMetaDataFields()) {
+                Object o = m.getMetaData(key);
+                if(o == null ) { continue;}
+                String value = o.toString();
+                values.put(TableBuilder.scrubName(key), value);
+            }
+        }
+        if(blob.length > 1000000){
+            throw new RecordTooLargeException(blob.length / 1000000);
+        }
+        values.put(DATA_COL, blob);
+
+        return values;
+    }
+
+    public static String getTableCreateString(String storageKey, Persistable p){
+        TableBuilder mTableBuilder = new TableBuilder(storageKey);
+        mTableBuilder.addData(p);
+        return mTableBuilder.getTableCreateString();
+    }
+
+    public static Pair<String, List<Object>> getTableInsertData(String storageKey, Persistable p){
+        TableBuilder mTableBuilder = new TableBuilder(storageKey);
+        mTableBuilder.addData(p);
+        return mTableBuilder.getTableInsertData(p);
+    }
+
+    public static String getRecordForValueSelectString(String storageKey, String id){
+        return "SELECT " + DatabaseHelper.DATA_COL + " from " + storageKey + " WHERE commcare_sql_id = " + id + ";";
```

Comment:
This looks injection vulnerable unless id _always_ comes from our code directly.